src/libcore/slice/mod.rs

```diff
@@ -394,23 +394,21 @@ impl<T> SliceExt for [T] {
     fn binary_search_by<'a, F>(&'a self, mut f: F) -> Result<usize, usize>
         where F: FnMut(&'a T) -> Ordering
     {
-        let mut base = 0usize;
-        let mut s = self;
-
-        loop {
-            let (head, tail) = s.split_at(s.len() >> 1);
-            if tail.is_empty() {
-                return Err(base)
-            }
-            match f(&tail[0]) {
-                Less => {
-                    base += head.len() + 1;
-                    s = &tail[1..];
-                }
-                Greater => s = head,
-                Equal => return Ok(base + head.len()),
-            }
+        let s = self;
+        let mut size = s.len();
+        if size == 0 {
+            return Err(0);
         }
+        let mut base = 0usize;
+        while size > 1 {
+            let half = size / 2;
+            let mid = base + half;
+            let cmp = f(&s[mid]);
```

Comment:
We should probably use unsafe getter here to avoid the bounds check. The performance gain is minimal but there's no good reason to retain it now that we're going for full speed.

Same for `&s[base]` bellow

Reply:
I think the compiler optimizes them out -- at least when I checked a version of this PR's code earlier.

Reply:
I can see them in this test https://godbolt.org/g/jTCpAi but I may have screwed something.

Reply:
I swear I tried this before and it had no effect. Now it seems that it does so I am patching this in. I am squashing all commits in a bit.